### PR TITLE
 * First pass at custom coerce for input objects

### PR DESCRIPTION
**Pull request by njlr**

### CODE REVIEW

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 1]
```diff
@@ -153,17 +158,29 @@ let rec private coerceVariableValue isNullable typedef (vardef: VarDef) (input:
...
+            coerceInput v |> Option.toObj
+        | _ -> failwithf "Expected an input Value but got %A" input
+    | None ->
+        match input with
+        | :? Map<string, obj> as map ->
+            let mapping =
+                objdef.Fields
+                |> Array.map (fun field ->
+                    let valueFound = Map.tryFind field.Name map |> Option.toObj
+                    let coercedValue = coerceVariableValue false field.TypeDef vardef valueFound (errMsg + (sprintf "in field '%s': " field.Name))
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs:175]:
```suggestion
                |> Seq.map (fun field ->
                    let valueFound = map |> Map.tryFind field.Name |> Option.toObj
                    let coercedValue = coerceVariableValue false field.TypeDef vardef valueFound (errMsg + (sprintf "in field '%s': " field.Name))
```
Maybe it does not worth to materialize it to an array

File: src/FSharp.Data.GraphQL.Shared/TypeSystem.fs [hunk 2]
```diff
@@ -1530,13 +1532,22 @@ and InputObjectDefinition<'Val> =
       Description : string option
       /// Function used to define field inputs. It must be lazy
       /// in order to support self-referencing types.
-      FieldsFn : unit -> InputFieldDef [] }
+      FieldsFn : unit -> InputFieldDef []
+      /// Optional function to customize default coercion behaviour
+      CoerceInput : (Value -> 'Val option) option }
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Shared/TypeSystem.fs:1537]: Have you considered using `Result<,>`?
Or @jberzy are there other places to move to `Result` so it is better to have a new PR for that?

**Review by xperiandri — COMMENTED** [review]
I like what you have done. This is useful.
Also, it would be more useful to be able to return an error and pass that error to the response with error to GraphQL field mapping.

File: src/FSharp.Data.GraphQL.Server/Values.fs [hunk 3]
```diff
@@ -38,31 +38,36 @@ let rec internal compileByType (errMsg: string) (inputDef: InputDef): ExecuteInp
     | Scalar scalardef ->
         variableOrElse (scalardef.CoerceInput >> Option.toObj)
```
Comment (xperiandri) [inline src/FSharp.Data.GraphQL.Server/Values.fs:39]: @jberzy can we switch this to `Result<,>` and pass an error to the response?

**Comment by xperiandri** [comment]
There is an issue I have with current scalar implementation
``` F#
    interface ScalarDef with
        member x.Name = x.Name
        member x.Description = x.Description
        member x.CoerceInput input = x.CoerceInput input |> Option.map box
        member x.CoerceValue value = (x.CoerceValue value) |> Option.map box
```

This piece is very contraintuitive
``` F#
        member x.CoerceInput input = x.CoerceInput input |> Option.map box
        member x.CoerceValue value = (x.CoerceValue value) |> Option.map box
```
@jberzy at least XML comments with clear description required.
Can we get rid of `CoerceValue` at all?

And `CoerceInput` here may be optional. Because if I preprocess variables with JSON deserializer I don't need any coercion at all.

File: src/FSharp.Data.GraphQL.Shared/TypeSystem.fs (hunk 2 above)
Comment (njlr) [reply src/FSharp.Data.GraphQL.Shared/TypeSystem.fs:1537]: I would also prefer result, although the current design is based around options (see coerce input methods)

**Comment by njlr** [comment]
@xperiandri I have fleshed out this idea further: https://github.com/fsprojects/FSharp.Data.GraphQL/pull/351

**Pull request closed without merging**